### PR TITLE
Initial code import

### PR DESCRIPTION
**Pull request by sbinq**

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	return Client{Transport: t, serverURL: serverURL}
+}
+
+// NewClientWithTimeout creates a client with http.Client
+func NewClientWithTimeout(serverURL string, timeout time.Duration) Client {
+	return NewClient(serverURL, NewDefaultTransport(timeout))
+}
+
+// Serve makes GET request to OSRM server and decodes the given JSON
+func (c Client) Serve(ctx context.Context, in *Request, out Response) error {
```
Comment (sbinq) [inline client.go]: Naming looks confusing here - `Serve` is usually associated with handling some requests - not making one.

File: match.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
+package osrm
+
+import geo "github.com/paulmach/go.geo"
+
+// MatchRequest represents a request to the match method
+type MatchRequest struct {
+	Request
```
Comment (sbinq) [inline match.go]: IMHO this `Request` embedding makes for not a really nice API from consumer side - you cannot write like
```go
Match(ctx, MatchRequest{
  Param1: v1,
  Param2: v2,
  ...
})
```
instead you'll have smth like
```go
Match(ctx, MatchRequest{
  Request: Request{
    Param1: v1,
  },
  Param2: v2,
  ...
})
```

File: options.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	opts[k] = append(opts[k], v...)
+}
+
+// Encode encodes the options into ``OSRM query`` form
+// ({option}={element};{element}[;{element} ... ]) sorted by key
+func (opts Options) Encode() string {
+	if opts == nil {
+		return ""
+	}
+	var buf []byte
```
Comment (sbinq) [inline options.go]: Maybe worth using `bytes.Buffer` here?

File: client.go (hunk 1 above)
Comment (regeda) [reply client.go]: Agree, `DoRequest` is appreciated.

File: match.go (hunk 2 above)
Comment (regeda) [reply match.go]: Agree, lets replace by own fields.

File: route.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	BearingAfter  float32   `json:"bearing_after"`
+	Location      geo.Point `json:"location"`
+	Type          string    `json:"type"`
+	Modifier      string    `json:"modifier"`
+}
+
+func (r RouteRequest) request() *Request {
+	r.service = "route"
+	r.options = stepsOptions(r.Steps, r.Annotations, r.Overview, r.Geometries)
+	r.options.Set("continue_straight", "true")
```
Comment (regeda) [inline route.go]: I guess that it shouldn't be hardcoded.

File: route.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	Bearings    []Bearing
+	Steps       Steps
+	Annotations Annotations
+	Overview    Overview
+	Geometries  Geometries
+}
+
+// RouteResponse represents a response from the route method
+type RouteResponse struct {
+	ResponseError
```
[the remark this answers is not in the document]
Comment (regeda) [reply route.go]: Let's make own fields.

File: types.go [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+	}, "/")
+	if len(r.options) > 0 {
+		url += "?" + r.options.Encode() // options
+	}
+	return url, nil
+}
+
+// Polyline generates a polyline in Google format
+// It uses default factor because of OSRM5 doesn't support polyline6 as coordinates
+func (g *GeoPath) Polyline() string {
```
Comment (regeda) [inline types.go]: Could you move functions near a definition?

File: osrm.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	defaultTimeout   = time.Second
+	defaultServerURL = "http://127.0.0.1:5000"
+
+	version = "v1"
+)
+
+// OSRM implements the common OSRM API v5.
+// See https://github.com/Project-OSRM/osrm-backend/blob/master/docs/http.md for details.
+// TODO: implement (nearest, trip, tile) methods
+type OSRM struct {
```
[the remark this answers is not in the document]
Comment (sbinq) [reply osrm.go]: Discussed, fine as is

File: client.go (hunk 1 above)
Comment (regeda) [reply client.go]: also, the function might be unexported

File: errors.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
+package osrm
+
+import "errors"
+
+type ErrCode string
```
[the remark this answers is not in the document]
Comment (sbinq) [reply errors.go]: removed

File: options.go [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+	keys := make([]string, 0, len(opts))
```
[the remark this answers is not in the document]
Comment (sbinq) [reply options.go]: `opts` is a map - so `i` type is `string` in your code; don't think it's worth adding separate variable with counter for this

File: options.go (hunk 3 above)
Comment (sbinq) [reply options.go]: Or maybe not

File: client.go (hunk 1 above)
Comment (sbinq) [reply client.go]: updated

File: match.go (hunk 2 above)
Comment (sbinq) [reply match.go]: updated

File: route.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sbinq) [reply route.go]: updated

File: route.go (hunk 4 above)
Comment (sbinq) [reply route.go]: updated

File: types.go (hunk 6 above)
Comment (sbinq) [reply types.go]: updated

File: osrm.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+// NewWithClient creates a client with custom transport layer
+func NewWithClient(client Client) *OSRM {
+	return &OSRM{client}
+}
+
+func (o OSRM) query(ctx context.Context, in *request, out response) error {
+	if err := o.client.doRequest(ctx, in, out); err != nil {
+		return err
+	}
+	if err := out.apiError(); err != nil {
```
Comment (regeda) [inline osrm.go]: just return
```
return out.apiError()
```

File: options.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+	var buf []byte
+	for _, k := range keys {
+		vals := opts[k]
+		if len(buf) > 0 {
+			buf = append(buf, '&')
+		}
+		buf = append(buf, url.QueryEscape(k)...)
+		buf = append(buf, '=')
+		for n, val := range vals {
```
Comment (regeda) [inline options.go]: you can omit `vals` variable and use `opts[k]` instead:
```
for n, val := range opts[k] {...}
```

File: route.go [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+		service: "route",
+		options: opts,
+	}
+}
+
+func stepsOptions(s Steps, a Annotations, o Overview, g Geometries) options {
+	opts := options{}
+
+	if steps := s.String(); steps != "" {
+		opts.Set("steps", steps)
```
Comment (regeda) [inline route.go]: I suggest to create the following func:
```
func (o options) SetStringer(k string, v Stringer) options {
    if s := v.String(); s != "" {
        o.Set(k, s)
    }
    return o
}
```
It helps to create a clear code:
```
opts := options{}
return opts.
    SetStringer("steps", s).
    SetStringer("annotations", a).
    SetStringer("geometries", g).
    SetStringer("overview", o)
```

File: options.go [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+	"net/url"
+	"sort"
+	"strconv"
+)
+
+// options represents OSRM query params to be encoded in URL
+type options map[string][]string
+
+// Set saves a string value by the key
+func (opts options) Set(k, v string) {
```
Comment (regeda) [inline options.go]: I think we can make all function unexported for `options` type.

File: README.md [hunk 14]
```diff
@@ -1,3 +1,60 @@
...
+	geo "github.com/paulmach/go.geo"
+)
+
+func main() {
+	client := osrm.NewFromURL("https://router.project-osrm.org")
+
+	ctx, cancelFn := context.WithTimeout(context.Background(), time.Second)
+	defer cancelFn()
+
+	resp, err := client.Route(ctx, osrm.RouteRequest{
```
[the remark this answers is not in the document]
Comment (sbinq) [reply README.md]: thanks, updated

File: options.go (hunk 13 above)
Comment (sbinq) [reply options.go]: updated

File: README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (sbinq) [reply README.md]: updated

File: options.go (hunk 13 above)
Comment (sbinq) [reply options.go]: updated

File: options.go (hunk 11 above)
Comment (sbinq) [reply options.go]: updated

File: route.go [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+	}
+
+	if annotations := a.String(); annotations != "" {
+		opts.Set("annotations", annotations)
+	}
+
+	opts.Set("geometries", GeometriesPolyline6.String())
```
[the remark this answers is not in the document]
Comment (sbinq) [reply route.go]: It is set below only if passed explicitly in request - here we set default value to be used of nothing passed explicitly

File: osrm.go (hunk 10 above)
Comment (sbinq) [reply osrm.go]: updated

File: route.go (hunk 12 above)
Comment (sbinq) [reply route.go]: updated

File: route.go [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+	Location      geo.Point `json:"location"`
+	Type          string    `json:"type"`
+	Modifier      string    `json:"modifier"`
+}
+
+func (r RouteRequest) request() *request {
+	opts := stepsOptions(r.Steps, r.Annotations, r.Overview, r.Geometries)
+
+	if cs := r.ContinueStraight.String(); cs != "" {
+		opts.set("continue_straight", cs)
```
Comment (regeda) [inline route.go]:
```go
opts := stepsOptions(r.Steps, r.Annotations, r.Overview, r.Geometries).
    setStringer("continue_straight", r.ContinueStraight)
```

File: route.go [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+
+func stepsOptions(steps Steps, annotations Annotations, overview Overview, geometries Geometries) options {
+	return options{}.
+		setStringer("steps", steps).
+		setStringer("annotations", annotations).
+		setStringer("geometries", valueOrDefault(geometries, GeometriesPolyline6)).
+		setStringer("overview", overview)
+}
+
+func valueOrDefault(geometries, def Geometries) Geometries {
```
Comment (regeda) [inline route.go]:
```go
func valueOrDefault(v, def fmt.Stringer) fmt.Stringer {
    if v.String() == "" {
       return def
    }
    return v
}
```

File: route.go (hunk 16 above)
Comment (sbinq) [reply route.go]: updated

File: route.go (hunk 17 above)
Comment (sbinq) [reply route.go]: updated

**Pull request merged**